### PR TITLE
optimize parquet write performance for parallel threads

### PR DESCRIPTION
**Pull request by binmahone**
### Changelog category (leave one):

- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

optimize parquet write performance for parallel threads

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

###  Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

In https://github.com/ClickHouse/ClickHouse/commit/eb3b774ad048e3cb6db34511950c75df27679739 , a Chunk buffer is introduced for staging small chunks until they accumulated to bigger than row group size. The cost of memory copy in the chunk buffer can be significant, especially when multiple threads are writing parquet files simultaneously.

This PR improves the chunk buffer by avoiding unnecessary memcpy.  My local experiments show a ~30% improvement for parallel writing with 31 threads on a 32-core CPU.

###  Experiments

```shell
for i in {10..40}                                                                                        
do
  nohup time /data0/hongbin/code/vanilla/ClickHouse/cmake-build-release/programs/clickhouse local "--query=INSERT INTO table function file('/data0/tmp/test_write2_$i.parquet',Parquet) select * from file('/data0/tpch100_zhichao/parquet/lineitem/part-000$i-*.parquet',Parquet)" > myoutput${i}.log 2>&1 &
done

cat myoutput* | grep "elapsed" | awk -F '[ :e]' '{sum+=$6}END{print sum/31}'
``` 
The above shell script first launches 31 threads to read&write 31 parquet files(each about 200+M) at the same time. After the threads are done, the last command collects the elapsed time and calculates the average time. The result is 12.7671 (this PR) vs. 17.8435(current master HEAD), showing a nearly 30% improvement

### CODE REVIEW

**Comment by binmahone** [comment]
hi @al13n321 , it's my first time to submit a PR to clickhouse, and I see many CI tests failing. The error message seems to be irrelevant to my PR, is it normal? Do I need to green light all the tests?

BTW, when I checked recent PRs, I find almost all PRs had a red light.

**Comment by binmahone** [comment]
hi @evillique @qoega, can you please give some advise for this PR? @al13n321  seems not responding

**Comment by evillique** [comment]
Builds fail ([report](https://s3.amazonaws.com/clickhouse-test-reports/50102/96dbc5aef871315ad1e64ed3f46b632753f80397/clickhouse_special_build_check/report.html)):
```
May 22 16:58:24 /build/src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:985:9: error: use range-based for loop instead [modernize-loop-convert,-warnings-as-errors]
May 22 16:58:24         for (size_t i = 0; i < chunks.size(); ++i)
May 22 16:58:24         ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
May 22 16:58:24             (const auto & chunk : chunks)


May 22 16:52:08 /build/src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp:117:57: error: no matching function for call to 'max'
May 22 16:52:08     auto status = file_writer->WriteTable(*arrow_table, std::max(format_settings.parquet.row_group_bytes, staging_bytes));
May 22 16:52:08                                                         ^~~~~~~~
May 22 16:52:08 /build/contrib/llvm-project/libcxx/include/__algorithm/max.h:40:1: note: candidate template ignored: deduced conflicting types for parameter '_Tp' ('UInt64' (aka 'unsigned long long') vs. 'size_t' (aka 'unsigned long'))
May 22 16:52:08 max(const _Tp& __a, const _Tp& __b)
May 22 16:52:08 ^
May 22 16:52:08 /build/contrib/llvm-project/libcxx/include/__algorithm/max.h:51:1: note: candidate template ignored: could not match 'initializer_list<_Tp>' against 'UInt64' (aka 'unsigned long long')
May 22 16:52:08 max(initializer_list<_Tp> __t, _Compare __comp)
May 22 16:52:08 ^
May 22 16:52:08 /build/contrib/llvm-project/libcxx/include/__algorithm/max.h:61:1: note: candidate function template not viable: requires single argument '__t', but 2 arguments were provided
May 22 16:52:08 max(initializer_list<_Tp> __t)
May 22 16:52:08 ^
May 22 16:52:08 /build/contrib/llvm-project/libcxx/include/__algorithm/max.h:31:1: note: candidate function template not viable: requires 3 arguments, but 2 were provided
May 22 16:52:08 max(const _Tp& __a, const _Tp& __b, _Compare __comp)
May 22 16:52:08 ^
May 22 16:52:08 1 error generated.
```

Stateless tests with parquet in the name also seem to fail:
- [Stateless tests (debug) [1/5]](https://s3.amazonaws.com/clickhouse-test-reports/50102/96dbc5aef871315ad1e64ed3f46b632753f80397/stateless_tests__debug__[1_5].html)
- [Stateless tests (debug) [3/5]](https://s3.amazonaws.com/clickhouse-test-reports/50102/96dbc5aef871315ad1e64ed3f46b632753f80397/stateless_tests__debug__[3_5].html)
- [Stateless tests (debug) [5/5]](https://s3.amazonaws.com/clickhouse-test-reports/50102/96dbc5aef871315ad1e64ed3f46b632753f80397/stateless_tests__debug__[5_5].html)

**Comment by evillique** [comment]
The overall status of the commit being red is not that important, you should look at the reports of the failed checks to check if they are related to your changes.

**Comment by al13n321** [comment]
> nearly 30% improvement

download

---

> optimize parquet write performance for parallel threads

Is it related to threads? Seems that this just reduces copying, which should help regardless of number of threads.

---

Looks good, just need to fix the empty table case and the CI failures (the ones that are not flaky):
 * `ClickHouse special build check` are real build errors/warnings on different compilers/platforms, please fix. Click through to the build log and search for " error:" to get the actual error messages. There are two:
```
May 29 11:21:48 /build/src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:985:14: error: 'auto &chunk' can be declared as 'const auto &chunk' [readability-qualified-auto,-warnings-as-errors]
May 29 11:21:48         for (auto & chunk : chunks)
May 29 11:21:48              ^
May 29 11:21:48              const 

May 29 11:19:16 /build/src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp:117:57: error: no matching function for call to 'max'
May 29 11:19:16     auto status = file_writer->WriteTable(*arrow_table, std::max(format_settings.parquet.row_group_bytes, staging_bytes));
May 29 11:19:16                                                         ^~~~~~~~
May 29 11:19:16 /build/contrib/llvm-project/libcxx/include/__algorithm/max.h:40:1: note: candidate template ignored: deduced conflicting types for parameter '_Tp' ('UInt64' (aka 'unsigned long long') vs. 'size_t' (aka 'unsigned long'))
May 29 11:19:16 max(const _Tp& __a, const _Tp& __b)
```
 * Not sure what's going on with `Performance Comparison`. Likely flaky. Let's see if it goes away when CI re-runs after updating the PR.
 * `01429_empty_arrow_and_parquet` - probably the empty table bug, see review comment. `00900_long_parquet_decimal`, `02511_parquet_orc_missing_columns`, `02566_ipv4_ipv6_binary_formats` - hopefully the same, they're all crashes, let's fix and see.
 * `02595_orc_arrow_parquet_more_types`, `02721_parquet_field_not_found` - should investigate, I don't think they're broken on master. https://s3.amazonaws.com/clickhouse-test-reports/50102/6abd894f28586e24fcbe14026927eb89cd11194c/stateless_tests__debug__[5_5].html

**Review by al13n321 — CHANGES_REQUESTED** [review]

**Comment by al13n321** [comment]
Ok, I see what's going on with the parquet tests. The `consume(header)` case was already broken, and this PR accidentally made this call happen even for nonempty tables. The review comments should be enough to make the tests pass.

Fix in https://github.com/ClickHouse/ClickHouse/pull/50328

**Comment by binmahone** [comment]
hi @al13n321 , yesterday when I checked your code change in PR https://github.com/ClickHouse/ClickHouse/pull/49367/files, I realized it is a superset of this one. Do you suggest me going ahead on this PR, or just wait for your PR to be merged?

**Comment by binmahone** [comment]
Is it related to threads? Seems that this just reduces copying, which should help regardless of number of threads -- you're right, but the gap seems to grow larger in multi-thread cases. The gap is neglectable under single thread case.

**Comment by al13n321** [comment]
> Do you suggest me going ahead on this PR, or just wait for your PR to be merged?

Up to you. I think this much simpler PR still makes sense, in case the other one ends up abandoned or takes a while to review.

> but the gap seems to grow larger in multi-thread cases. The gap is neglectable under single thread case.

Interesting! I haven't seen (or haven't noticed) this effect before. Maybe it gets bottlenecked on memory bandwidth when processing 31x more total throughput. TIL to worry more about unnecessary memory accesses when there are more threads.

**Comment by binmahone** [comment]
if you're interested in multi-thread cases, here's more information:

Per my experiments on a server with intel 11900K CPU (8 cores + 8 hyper thread cores = 16 total vcores), performance grows linearly from 1 thread to 4 threads. But the performance improvement from 4 cores to 8 cores is very minor. This surprised me because we've just reached half the total cores.

I did some profiling with vtune, showing memory bound being more significant from 4 thread to 8 thread. However, most of the hotspot functions are Arrow codes and I have no clue how to optimize them.

File: src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp
Comment (binmahone) [reply src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp:149]: changed

File: src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp
Comment (binmahone) [reply src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp:142]: changed

File: src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp
Comment (binmahone) [reply src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp:110]: Wait, doesn't WriteTable() take number of rows, not bytes?

you're right, I mixed them up.

Why max?

this is meant to replace https://github.com/ClickHouse/ClickHouse/blob/eb3b774ad048e3cb6db34511950c75df27679739/src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp#L94


Would the TODO amount to just copy-pasting the num_row_groups and row_group_size lines from before the PR?

I'm not sure what does the TODO comment mean, I just kept it untouched

File: src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp
Comment (binmahone) [reply src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:1038]: updated

File: src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp
Comment (al13n321) [reply src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp]: > I'm not sure what does the TODO comment mean, I just kept it untouched

The TODO is added by this very PR, check [the commit](https://github.com/ClickHouse/ClickHouse/pull/50102/files) :) . It's probably an incorrectly resolved merge conflict: the TODO used to exist, then was removed in https://github.com/ClickHouse/ClickHouse/pull/49325 , then this PR is inadvertently bringing it back. You should probably un-remove the following code to replace the TODO:
```
    /// Increase row group size slightly (by < 2x) to avoid adding a small row groups for the
    /// remainder of the new chunk.
    /// E.g. suppose input chunks are 70K rows each, and max_rows = 1M. Then we'll have
    /// getNumRows() = 1.05M. We want to write all 1.05M as one row group instead of 1M and 0.05M.
    size_t num_row_groups = std::max(static_cast<UInt64>(1), accumulated_chunk.getNumRows() / target_rows);
    size_t row_group_size = (accumulated_chunk.getNumRows() - 1) / num_row_groups + 1; // round up
```

File: src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp
Comment (binmahone) [reply src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:1039]: hi @al13n321 , I didn't quite get it. Is there any difference between your suggested code and mine?

Comment (al13n321) [reply src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:1039]: You're assigning `arrow_schema` only if it's the first `chChunkToArrowTable()` call on this `CHColumnToArrowColumn` instance.

Comment (binmahone) [reply src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:1039]: yes, that's my intention. and I'm checking the is_arrow_fields_initialized flat to ensure this. (while you're checking arrow_schema being null), I'm not sure what's the problem with my solution ?

File: src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp
Comment (binmahone) [reply src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp]: the removed code is meant to "avoid adding a small row groups for the remainder of the new chunk.", IMHO this can be replaced by https://github.com/ClickHouse/ClickHouse/blob/565575cef9e5032d9424c8f649fa94a731bf8779/src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp#L118

File: src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp
Comment (al13n321) [reply src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:1039]: `arrow_schema` is then passed to `arrow::Table::Make()` (below), which dereferences it and crashes if it's null.

Comment (binmahone) [reply src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp]: my bad..  I didn't realize arrow_schema is a local variable.. 
code updated. please check if there's any other issues with this PR

**Review by al13n321 — APPROVED** [review]

**Comment by al13n321** [comment]
Looks good, the remaining failed test is flaky (01154_move_partition_long).

**Pull request merged**
